docs/contributors/coding-guidelines.md

```diff
@@ -201,6 +201,246 @@ alert( 'My name is ' + name + '.' );
 alert( `My name is ${ name }.` );
 ```
 
+## JavaScript Documentation using JSDoc
+
+Gutenberg follows the [WordPress JavaScript Documentation Standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/), with additional guidelines relevant for its distinct use of [import semantics](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#imports) in organizing files, the [use of TypeScript tooling](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/testing-overview.md#javascript-testing) for types validation, and automated documentation generation using [`@wordpress/docgen`](https://github.com/WordPress/gutenberg/tree/master/packages/docgen).
+
+For additional guidance, consult the following resources:
+
+- [JSDoc Official Documentation](https://jsdoc.app/index.html)
+- [TypeScript Supported JSDoc](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#supported-jsdoc)
+
+### Custom Types
+
+Define custom types using the [JSDoc `@typedef` tag](https://jsdoc.app/tags-typedef.html).
+
+A custom type should include a description, and should always include its base type.
+
+Custom types should be named as succinctly as possible, while still retaining clarity of meaning and avoiding conflict with other global or scoped types. A `WP` prefix should be applied to all custom types. Avoid superfluous or redundant prefixes and suffixes (for example, a `Type` suffix, or `Custom` prefix). Custom types are not global by default, so a custom type does not need to be excessively specific to a particular package. However, they should be named with enough specificity to avoid ambiguity or name collisions when brought into the same scope as another type.
+
+```js
+/**
+ * A block selection object.
+ *
+ * @typedef {Object} WPBlockSelection
+ *
+ * @property {string} clientId     A block client ID.
+ * @property {string} attributeKey A block attribute key.
+ * @property {number} offset       An attribute value offset, based on the rich
+ *                                 text value.
+ */
+```
+
+Custom types can also be beneficial when describing a set of predefined options. While the [type union](https://jsdoc.app/tags-type.html) can be used with literal values as an inline type, using it in a custom type can afford the opportunity to describe the purpose of these options, and avoids a situation where aligning tags while still respecting a maximum line length of 80 characters becomes challenging due to the space occupied by a literal values union.
```

Comment:
```suggestion
While the [type union](https://jsdoc.app/tags-type.html) can be used with literal values as an inline type, it might quickly become challenging to align tags while still respecting a maximum line length of 80 characters. Using it in a custom type can afford the opportunity to describe the purpose of these options, and avoids the issue created by the space occupied by a literal values union.
```

I'd be nice to split this sentence into smaller parts to make it easier to read. I don't know if my proposal makes it any better but it shows the idea.

Reply:
>I'd be nice to split this sentence into smaller parts to make it easier to read. I don't know if my proposal makes it any better but it shows the idea.

I went with something slightly different in 1bb94137d, but hopefully to the same effect.